src/main/java/net/sf/javaanpr/recognizer/KnnPatternClassificator.java

```diff
@@ -60,7 +60,7 @@ public RecognizedChar recognize(Char chr) {
         RecognizedChar recognized = new RecognizedChar();
         for (int x = 0; x < this.learnVectors.size(); x++) {
             float fx = this.simplifiedEuclideanDistance(tested, this.learnVectors.elementAt(x));
-            recognized.addPattern(recognized.new RecognizedPattern(ALPHABET[x], fx));
+            recognized.addPattern(new RecognizedChar.RecognizedPattern(ALPHABET[x], fx));
```

Comment:
Optional: The way the `RecognizedPattern`, `RecognizedChar` and `PatternComparer` classes are nested in `CharacterRecognizer` is a pain. It's good that you made them static, but I would suggest going even further and pulling them out into separate files.